### PR TITLE
Develop modify stepping logic

### PR DESCRIPTION
**Pull request by vincentpierre**
### Proposed change(s)

Making changes to the step counting in both the Academy and the Agent.
The Academy will first set status, then tick its step count, then tick all agents step counts.
Collect observations and Act happen as usual.
At the end of the loop, the Agent resets if max step was reached.

### Useful links (Github issues, JIRA tickets, ML-Agents forum threads etc.)

Is continuation of #3432 
Jira is MLA-619

### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [x] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added updated the changelog (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] I have updated the migration guide (if applicable)

### Other comments

### CODE REVIEW

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 1]
```diff
@@ -688,19 +690,16 @@ void SendInfo()
             }
         }
 
+        void AgentIncrementStep()
+        {
+
+            m_StepCount += 1;
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Agent.cs]: nit: extra whitespace above and below.

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 2]
```diff
@@ -688,19 +690,16 @@ void SendInfo()
             }
         }
 
+        void AgentIncrementStep()
+        {
+
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Agent.cs]:
```suggestion
```

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 3]
```diff
@@ -688,19 +690,16 @@ void SendInfo()
...
+            m_StepCount += 1;
+
```
Comment (surfnerd) [inline com.unity.ml-agents/Runtime/Agent.cs]:
```suggestion
```

**Comment by surfnerd** [comment]
what does the step count look like for an episode?

It looks like if `maxStep = 5` the step count will look like:
```
AgentStep = 1
  Agent Observe; Agent Act;
AgentStep = 2
  Agent Observe; Agent Act;
AgentStep = 3
  Agent Observe; Agent Act;
AgentStep = 4
  Agent Observe; Agent Act;
AgentStep = 5
  Agent Observe; Agent Act;
Agent Reset;
```

Is my understanding correct?

**Comment by vincentpierre** [comment]
> Is my understanding correct?

Yes, this is what I am aiming at

File: com.unity.ml-agents/Runtime/Agent.cs [hunk 4]
```diff
@@ -258,6 +259,7 @@ void OnDisable()
             // We don't want to even try, because this will lazily create a new Academy!
             if (Academy.IsInitialized)
             {
+                Academy.Instance.AgentIncrementStep -=AgentIncrementStep;
```
Comment (chriselion) [inline com.unity.ml-agents/Runtime/Agent.cs]:
```suggestion
                Academy.Instance.AgentIncrementStep -= AgentIncrementStep;
```

File: com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs [hunk 5]
```diff
@@ -524,12 +525,11 @@ public void TestMaxStepsReset()
 
             for (var i = 0; i < 15; i++)
             {
-                // We expect resets to occur when there are maxSteps actions since the last reset (and on the first step)
-                var expectReset = agent1.agentActionCallsSinceLastReset == maxStep || (i == 0);
+                Debug.Log(i);
+                // We expect resets to occur when there are maxSteps actions since the last reset 
+                var expectReset = agent1.agentActionCallsSinceLastReset == maxStep;
```
Comment (chriselion) [inline com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: Hang on, this test looks broken. `var previousNumResets` used to be the number of resets before stepping. But now we assert on its value before stepping the Academy. So `previousNumResets + 1, agent1.agentResetCalls` can never be true, which means expectReset must never be true. So you're basically testing `agent1.agentResetCalls== agent1.agentResetCalls1`

Comment (vincentpierre) [reply com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: Re-wrote the test that was broken

Comment (chriselion) [reply com.unity.ml-agents/Tests/Editor/MLAgentsEditModeTest.cs]: Looks pretty good. I made a few additions on https://github.com/Unity-Technologies/ml-agents/pull/3467 - can you merge that if it seems reasonable?

**Pull request merged**